Wip 18533 tool

dzafman opened the pull request:
Improve DBObjectMap testing
Improvements for ceph_osdomap_tool
Add ceph_test_object_map to make check
Simplify rm_keys() by eliminating the use of complete mappings and just copy on clone

dzafman commented:
@athanatos If this is merged before your branch, then Jenkins testing for your DBOjbectMap fixes will be meaningful.

dzafman commented:
retest this please.

dzafman commented:
@athanatos @jdurgin I will just copy all keys on clone and add a rename.  I will keep the commit from Sam in case we ever wanted to revert the clone/rename change.

athanatos commented:
@dzafman You have to deal with existing objects which are children of parents even if you don't create new ones, so you still need to fix the bug in rmkeys.

dzafman commented:
@athanatos Conceivably, at the start of rm_keys I could replace the full set of keys, remove the complete mapping and then perform the removal.

dzafman commented:
@jdurgin I tried to make the whole DBObjectMap code easier to work with.  We are moving to BlueStore so I'm less concern about extra clone overhead for filestore.  I'm leaving @athanatos commits that fix rm_keys() in case we wanted to revert d1b021d55e6d39f0cbc23957325e90a7dd1fb901 .  Another idea would be to take a hybrid approach and keep clone fast, but require the extra overhead in rm_keys() by copying and breaking parent connection.

dmick commented on src/tools/ceph_osdomap_tool.cc:
```diff
@@ -161,6 +161,7 @@ int main(int argc, char **argv) {
     r = omap.check(std::cout);
```
commit msg has typo "ceoh"

dzafman commented:
Testing passed:

dzafman-2017-03-03_16:32:03-rados-wip-18533-tool-distro-basic-smithi
879980 BUG #15912 Bluestore ENOSPC - already fixed
880000 BUG #18929 - already fixed
880012 BUG #19099 valgrind OSD slow start - already fixed

dzafman commented:
I renamed old clone call to legacy_clone() so my commit that we could reverse is c92c9f26180ad68c1b66137d51bd95878fe8136c if we want the original clone performance.

jdurgin commented on src/tools/ceph_osdomap_tool.cc:
```diff
@@ -27,12 +27,13 @@ using namespace std;
     ("omap-path", po::value<string>(&store_path),
      "path to mon directory, mandatory (current/omap usually)")
     ("paranoid", "use paranoid checking")
+    ("oid", po::value<string>(&oid), "Restrict to this object id")
```
add ' when dumping objects' since it only applies to dump-objects and dump-objects-with-keys

jdurgin commented on src/os/filestore/DBObjectMap.h:
```diff
@@ -276,7 +276,7 @@ class DBObjectMap : public ObjectMap {
     uint64_t num_children;
 
-    coll_t c;
+    coll_t c; // Not used
```
could remove and use a dummy coll_t in encode/decode

jdurgin commented on src/os/filestore/DBObjectMap.cc:
```diff
@@ -78,18 +78,24 @@ int DBObjectMap::check(std::ostream &out)
+          t->rmkeys_by_prefix(USER_PREFIX + header_key(header.seq) + COMPLETE_PREFIX);
+          db->submit_transaction(t);
+          out << "Cleared complete mapping to repair" << std::endl;
+        } else
```
need braces for the else clause - easy to mistake it for part of the loop otherwise

jdurgin commented on src/tools/ceph_osdomap_tool.cc:
```diff
@@ -157,14 +160,16 @@ int main(int argc, char **argv) {
-  } else if (cmd == "check") {
-    r = omap.check(std::cout);
+  } else if (cmd == "check" || cmd == "repair") {
+    bool repair = false;
```
repair = (cmd == "repair")

jdurgin commented on src/tools/ceph_osdomap_tool.cc:
```diff
@@ -180,6 +180,12 @@ int main(int argc, char **argv) {
     }
     for (auto i : headers)
       std::cout << i << std::endl;
+  } else if (cmd == "corrupt") {
```
where is this used? do you have a test script using it that isn't in this branch?

jdurgin replied:
it would be helpful to have a test that uses this, and checks that the repair op works on the corrupted store

dzafman replied:
I haven't added testing of the "corrupt" command of ceph-osdomap-tool.  It doesn't mean anything after my last commit because we no longer use complete mappings.

jdurgin commented on src/os/filestore/DBObjectMap.cc:
```diff
@@ -78,18 +78,24 @@ int DBObjectMap::check(std::ostream &out)
+          KeyValueDB::Transaction t = db->get_transaction();
+          t->rmkeys_by_prefix(USER_PREFIX + header_key(header.seq) + COMPLETE_PREFIX);
+          db->submit_transaction(t);
+          out << "Cleared complete mapping to repair" << std::endl;
```
the commit message should have an explanation of why clearing the complete mapping is the right way to repair this:
- why do we not need to remove the reference to the parent?
- why will queries be correct after this?
- what are our assumptions about the state of the keyvaluedb here?

dzafman replied:
@jdurgin I added this to the commit comment for the repair:

Under some situations the previous rm_keys() code would
generated a corrupt complete table.  There is no way to
figure out what the table should look like now.  By removing
the entries we fix the corruption and aren't much worse off
because the corruption caused some deleted keys to re-appear.

This doesn't breaking the parent/child relationship during
repair because some of the keys may still be contained
in the parent.

jdurgin commented on src/os/filestore/DBObjectMap.cc:
```diff
@@ -1375,3 +1379,42 @@ int DBObjectMap::corrupt()
+    if (destination) {
+      if (check_spos(to, destination, spos))
+	return 0;
+      return -EEXIST;
```
returning -EEXIST in this case differs from clone - can the caller in filestore handle that? is that case possible?

dzafman replied:
I put the same code here as in clone() when the target already exists.  I also added to a test case for this code path.

dzafman commented:
@jdurgin I addressed code review comments and rebased to get @liewegas's new full handling code.  I changed the injectfull code to set cur_state so that the mon will notice and update status information.

dzafman commented:
Still need to squash commits

jdurgin commented:
Fixups look good, thanks! Since the corrupt command does nothing after the last commit, seems like we should just drop it.

dzafman commented:
@jdurgin I remove the undocumented "corrupt" command in the ceph-osdomap-tool.  Also, squashed commits and will start rados suite.

dzafman commented:
Tests passed:

dzafman-2017-03-09_12:29:18-rados-wip-18533-tool-distro-basic-smith
898763 BUG 18926   
898781 BUG 18993 (already fixed)
898854 BUG 18664 (new opened)
898899 Installation issue (test infrastructure)
898949 BUG ????? valgrind slow osd start
898981 DEAD Test issue BUG 10902

dzafman commented:
@jdurgin @dmick Should we try this on the long-running cluster?  I just realized I want to do an cephfs test run before merge.

jdurgin commented:
a cephfs test run sounds good

tchaikov commented:
i will run it with the rados qa suites.

dzafman commented:
@tchaikov Already passed rados run last week.

@jcsp Can you look at http://pulpito.ceph.com/dzafman-2017-03-10_11:31:53-fs-wip-18533-tool-distro-basic-smithi/ to verify that the cephfs failures/dead jobs are not related to this change?  The issue that is being fixed are the omaps which affect metadata pool directories I think if omap becomes corrupted as it did on LRC.

jcsp commented:
> Dead: 2017-03-11T07:38:24.224 DEBUG:teuthology.exit:Finished running handlers
> 2 jobs: ['901691', '901723']

Fuse mount failure with valgrind.  This is fixed in more recent master (http://tracker.ceph.com/issues/18797)

> Failure: reached maximum tries (50) after waiting for 300 seconds
> 2 jobs: ['901843', '901816']

Time out waiting for OSDs healthy on start.  Don't know why.

> Valgrind: osd (InvalidRead)
> 1 jobs: ['901784']

```
<kind>InvalidRead</kind>
  <what>Invalid read of size 8</what>
  <stack>
    <frame>
      <ip>0xB35F59</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>cmp(hobject_t const&amp;, hobject_t const&amp;)</fn>
      <dir>/usr/src/debug/ceph-12.0.0-292-g36b27fe/src/common</dir>
      <file>hobject.cc</file>
      <line>328</line>
    </frame>
    <frame>
      <ip>0x61609A</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>operator&gt;</fn>
      <dir>/usr/src/debug/ceph-12.0.0-292-g36b27fe/src/common</dir>
      <file>hobject.h</file>
      <line>294</line>
    </frame>
    <frame>
      <ip>0x61609A</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>have_backoff</fn>
      <dir>/usr/src/debug/ceph-12.0.0-292-g36b27fe/src/osd</dir>
      <file>Session.h</file>
      <line>172</line>
    </frame>
    <frame>
      <ip>0x61609A</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>PG::add_backoff(boost::intrusive_ptr&lt;Session&gt;, hobject_t const&amp;, hobject_t const&amp;)</fn>
      <dir>/usr/src/debug/ceph-12.0.0-292-g36b27fe/src/osd</dir>
      <file>PG.cc</file>
      <line>2329</line>
    </frame>
    <frame>
      <ip>0x6FF689</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>PrimaryLogPG::do_op(boost::intrusive_ptr&lt;OpRequest&gt;&amp;)</fn>
      <dir>/usr/src/debug/ceph-12.0.0-292-g36b27fe/src/osd</dir>
      <file>PrimaryLogPG.cc</file>
      <line>1941</line>
    </frame>
    <frame>
      <ip>0x6BF0F0</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>PrimaryLogPG::do_request(boost::intrusive_ptr&lt;OpRequest&gt;&amp;, ThreadPool::TPHandle&amp;)</fn>
      <dir>/usr/src/debug/ceph-12.0.0-292-g36b27fe/src/osd</dir>
      <file>PrimaryLogPG.cc</file>
      <line>1698</line>
    </frame>
```

Seems pretty unlikely any of the failures are from this PR.

liewegas commented:
@jcsp path for that valgrind failure?

jdurgin commented:
@liewegas http://qa-proxy.ceph.com/teuthology/dzafman-2017-03-10_11:31:53-fs-wip-18533-tool-distro-basic-smithi/901784/remote/smithi039/log/valgrind/osd.2.log.gz

liewegas commented:
ok cool, looks like this bug was already fixed!

tchaikov commented:
@dzafman ack, the PR passed my rados qa run also.

dzafman commented:
@tchaikov Please review on Monday.  If you have questions maybe we can discuss Monday night (my morning).

dzafman commented:
@dmick Please indicate if the LRC is running stable with the 11.2.0 version of this change.

dmick commented:
No problems so far

dzafman commented:
@jdurgin Rebased due to new conflicts.  This has had only a review by you.  It seems to be ok running on the LRC.  Should we merge it if it passes Jenkins run?
